spirv_cross.cpp

```diff
@@ -4910,11 +4910,13 @@ void Compiler::make_constant_null(uint32_t id, uint32_t type)
 		uint32_t parent_id = ir.increase_bound_by(1);
 		make_constant_null(parent_id, constant_type.parent_type);
 
-		if (!constant_type.array_size_literal.back())
-			SPIRV_CROSS_THROW("Array size of OpConstantNull must be a literal.");
-
-		SmallVector<uint32_t> elements(constant_type.array.back());
-		for (uint32_t i = 0; i < constant_type.array.back(); i++)
+		// The array size of OpConstantNull can be either literal or specialization constant.
+		// In the latter case, we cannot take the value as-is, as it can be changed to anything.
+		// Rather, we assume it to be *one* for the sake of initializer.
+		bool is_literal_array_size = constant_type.array_size_literal.back();
+		uint32_t count = is_literal_array_size ? constant_type.array.back() : 1;
```

Comment:
Does this work only because we kinda throw away the array size of an OpConstantNull later when initializing the threadgroup variable anyway?

Reply:
It is the case for shared (threadgroup) variable, but not for other kinds. In those cases, the initializer list is still used. In fact, I'm not 100% sure if the handling is correct for those cases where initializer is used (although I hope it is if the languages act like C/C++). At least Apple's MSL compiler does not complain about it. For GLSL, I'm not an expert to say if it is right. We could again use `GL_EXT_null_initializer` to eliminate doubt but it seems to require some extra work I don't quite know how. I tried to set the array size to 0, but it seems to turn the array into a scalar value.

Reply:
I discovered the logic this part is a little problematic for glsl. I'm adding a test and trying to resolve it.

Reply:
Apparently GLSL doesn't like it when initializer and the corresponding array having different number of elements, which will *certainly* happen for (non-shared) null constant array of specialized dimensions. Although it is a rare case, I decided to try my best to generate valid code with the least level of invasive change. I decided to only apply `GL_EXT_null_initializer` on `OpConstantNull` arrays with non-literal length (tracked with a flag at creation), and it is limited to the GLSL backend. A glsl test is also introduced.